src/main/java/de/hysky/skyblocker/skyblock/item/tooltip/adders/CraftPriceTooltip.java

```diff
@@ -0,0 +1,96 @@
+package de.hysky.skyblocker.skyblock.item.tooltip.adders;
+
+import de.hysky.skyblocker.config.SkyblockerConfigManager;
+import de.hysky.skyblocker.config.configs.GeneralConfig;
+import de.hysky.skyblocker.skyblock.item.tooltip.ItemTooltip;
+import de.hysky.skyblocker.skyblock.item.tooltip.TooltipAdder;
+import de.hysky.skyblocker.skyblock.item.tooltip.TooltipInfoType;
+import de.hysky.skyblocker.utils.NEURepoManager;
+import io.github.moulberry.repo.data.NEUIngredient;
+import io.github.moulberry.repo.data.NEUItem;
+import io.github.moulberry.repo.data.NEURecipe;
+import net.minecraft.item.ItemStack;
+import net.minecraft.screen.slot.Slot;
+import net.minecraft.text.Text;
+import net.minecraft.util.Formatting;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class CraftPriceTooltip extends TooltipAdder {
+    private final Map<String, Double> cachedCraftCosts = new HashMap<>();
+    private static final int MAX_RECURSION_DEPTH = 15;
+
+    public CraftPriceTooltip(int priority) {
+        super(priority);
+    }
+
+    @Override
+    public void addToTooltip(@Nullable Slot focusedSlot, ItemStack stack, List<Text> lines) {
+        if (SkyblockerConfigManager.get().general.itemTooltip.enableCraftingCost == GeneralConfig.Craft.OFF) return;
+
+        String internalID = stack.getSkyblockId();
+        if (stack.getNeuName() == null || internalID == null) return;
+
+        if (TooltipInfoType.LOWEST_BINS.getData() == null || TooltipInfoType.BAZAAR.getData() == null) {
+            ItemTooltip.nullWarning();
+            return;
+        }
+
+        NEUItem neuItem = NEURepoManager.NEU_REPO.getItems().getItemBySkyblockId(internalID);
+        if (neuItem == null) return;
+
+        List<NEURecipe> neuRecipes = neuItem.getRecipes();
+        if (neuRecipes.isEmpty() || neuRecipes.getFirst().getClass().equals(io.github.moulberry.repo.data.NEUKatUpgradeRecipe.class)) return;
+
+        double totalCraftCost = getItemCost(neuRecipes.getFirst(), 0);
+        cachedCraftCosts.clear();
```

Comment:
Why do you clear the cache right after getting the item cost? Also the cache is not checked when displaying the tooltip it seems? Why does it exist??

Reply:
you could make it so the cache gets cleared when the config changes, you can add a listener to an option in yacl

Reply:
I think it's 9^15 calculations at the worst case (could be 9^14 or 9^16, didn't check for an off-by-one), which is around 205 billion calculations. The cache limits that number to the number of items that exist in the game at worst case, which is pretty neat.

I think some kind of cache would be useful though, like a `Map<String, List<Pair<String, Int>>>` (or something similar with fastutil's classes to avoid the unnecessary boxing and unboxing) that maps each skyblock item id to a list of pairs with item ids and their amounts used in crafting recipe according to the given recursion limit, and gets reset when the recursion depth is changed. Then you could just get the prices of the items in the list for each item and multiply them by their amounts to figure out the cost. This also prevents the issue of price changes, as you calculate it dynamically.